File: backend/services/gemini.js
const axios = require('axios');

const GEMINI_BASE = 'https://generativelanguage.googleapis.com/v1beta/models';

const ROOTS_SYSTEM_PROMPT = `You are Roots 🌱, a warm and knowledgeable AI companion for newcomers to Canada.
You help immigrants navigate settlement — from documents and healthcare to career and community.

CRITICAL: ALWAYS respond in the exact same language the user uses (French, Spanish, Arabic, etc.).
Be concise (1-3 short paragraphs max), practical, and empathetic.
When relevant, mention: SIN, health card, banking, housing, job search, language classes, or community resources.
If you don't know something specific, say so and suggest calling 211 (Canada's social services helpline).`;

/**
 * Generate a structured critical path of tasks for a newcomer profile
 */
async function generateCriticalPath(profile) {
  const url = `${GEMINI_BASE}/gemini-2.5-flash:generateContent?key=${process.env.GEMINI_API_KEY}`;
  const prompt = `You are an expert on Canadian immigration and newcomer onboarding.
Given this immigrant profile, generate a sequenced critical path of tasks for their first 90 days.
Return JSON only. No markdown. No backticks. No explanation.

Profile: ${JSON.stringify(profile)}

Return exactly this structure with 5-7 tasks:
{"tasks":[{"id":"sin","title":"Apply for your SIN","description":"Your Social Insurance Number is required before you can work legally in Canada.","daysFromArrival":1,"urgency":"critical","estimatedTime":"2-3 hours"}]}`;

  const response = await axios.post(url, {
    contents: [{ parts: [{ text: prompt }] }]
  });

  const text = response.data.candidates[0].content.parts[0].text;
  return JSON.parse(text.replace(/```json|```/g, '').trim());
}

/**
 * General-purpose Gemini chat with conversation history and tool support.
 * Used as the brain for all active Twilio/WhatsApp conversations.
 * @param {string} message - The user's latest message
 * @param {Array} history - Array of { role: 'user'|'model', text: string }
 * @param {string} [systemPrompt] - Optional override system prompt
 * @param {string} [userPhoneNumber] - The user's phone number for outbound calls
 */
async function chat(message, history = [], systemPrompt = null, userPhoneNumber = null) {
  const url = `${GEMINI_BASE}/gemini-2.5-flash:generateContent?key=${process.env.GEMINI_API_KEY}`;

  const systemInstruction = {
    parts: [{ text: systemPrompt || ROOTS_SYSTEM_PROMPT }]
  };

  // Build conversation contents from history
  const contents = [];
  for (const turn of history) {
    contents.push({
      role: turn.role,
      parts: [{ text: turn.text }]
    });
  }

  // Add the current user message
  contents.push({
    role: 'user',
    parts: [{ text: message }]
  });

  const tools = [{
    functionDeclarations: [{
      name: "makeOutboundCall",
      description: "Triggers an AI phone assistant to call the user immediately. Use this ONLY when the user explicitly asks to be called, speak on the phone, or have a phone conversation.",
      parameters: {
        type: "OBJECT",
        properties: {
          reason: {
            type: "STRING",
            description: "A short reason for the call"
          }
        },
        required: ["reason"]
      }
    }]
  }];

  const payload = {
    systemInstruction,
    contents,
    tools
  };

  const response = await axios.post(url, payload);
  const part = response.data.candidates[0].content.parts[0];

  // Handle function call (user requested a phone call)
  if (part.functionCall && part.functionCall.name === 'makeOutboundCall') {
    // Instead of initiating an outbound call, instruct the user to call inbound
    const formattedNumber = process.env.VAPI_PHONE_NUMBER_ID || process.env.TWILIO_WHATSAPP_FROM.replace('whatsapp:', '');
    return `I can't make outbound calls right now, but you can call me directly at ${formattedNumber}! Just use your phone to call me and we can talk 📞`;
  }

  // Standard text response
  return part.text ? part.text.trim() : "I'm here to help!";
}

/**
 * Transcribe audio using Gemini's multimodal capabilities
 * @param {Buffer} audioBuffer
 * @param {string} mimeType
 */
async function transcribeAudio(audioBuffer, mimeType = 'audio/webm') {
  const url = `${GEMINI_BASE}/gemini-2.5-flash:generateContent?key=${process.env.GEMINI_API_KEY}`;
  const base64Audio = audioBuffer.toString('base64');

  const response = await axios.post(url, {
    contents: [{
      parts: [
        { text: 'Transcribe this audio accurately. Return only the transcription text, nothing else.' },
        { inlineData: { mimeType, data: base64Audio } }
      ]
    }]
  });

  return response.data.candidates[0].content.parts[0].text.trim();
}

/**
 * Generates a dynamically phrased onboarding question based on previous context.
 * @param {Array} previousAnswers - Array of previous questions and answers
 * @param {string} targetTopic - The core question to ask next
 */
async function generateNextQuestion(previousAnswers, targetTopic) {
  const url = `${GEMINI_BASE}/gemini-2.5-flash:generateContent?key=${process.env.GEMINI_API_KEY}`;

  let context = "";
  if (previousAnswers && previousAnswers.length > 0) {
    context = `Here is what we know about the user so far:\n${previousAnswers.join('\n')}\n\n`;
  }

  const prompt = `You are Roots 🌱, a warm AI companion for newcomers to Canada.
You are currently onboarding a new user.
${context}
It is your turn to ask the user a question. The core information you need to gather is:
"${targetTopic}"

Rephrase this core question into a single, naturally conversational and friendly message.
You can briefly acknowledge their previous answer if it makes sense, but keep it short.
DO NOT provide any advice yet, just ask the question.
Your entire response must be ONLY the question (1-2 sentences max).`;

  const response = await axios.post(url, {
    contents: [{ parts: [{ text: prompt }] }]
  });

  return response.data.candidates[0].content.parts[0].text.trim();
}

module.exports = { generateCriticalPath, chat, transcribeAudio, generateNextQuestion };

File: backend_py/app/services/gemini.py
"""Gemini calls used outside the ADK agents.

The ADK agents drive their own Gemini conversation, but a handful of
non-agent flows (critical-path generation, conversational onboarding
phrasing, raw multimodal extraction) still need direct access.
"""
from __future__ import annotations

import base64
import json
import logging
import re
from typing import Any

import httpx

from app.config import settings

log = logging.getLogger(__name__)

_API_BASE = "https://generativelanguage.googleapis.com/v1beta/models"


def _api_key() -> str:
    import os

    key = os.getenv("GEMINI_API_KEY", "")
    if not key:
        raise RuntimeError("GEMINI_API_KEY not configured")
    return key


def _strip_fences(text: str) -> str:
    return re.sub(r"^```(?:json)?|```$", "", text.strip(), flags=re.MULTILINE).strip()


async def generate_critical_path(profile: dict[str, Any]) -> dict[str, Any]:
    """Return ``{"tasks": [...]}`` — same shape the JS coordinator expects."""
    prompt = (
        "You are an expert on Canadian immigration and newcomer onboarding.\n"
        "Given this immigrant profile, generate a sequenced critical path of "
        "tasks for their first 90 days.\n"
        "Return JSON only. No markdown. No backticks. No explanation.\n\n"
        f"Profile: {json.dumps(profile, default=str)}\n\n"
        "Return exactly this structure with 5-7 tasks: "
        '{"tasks":[{"id":"sin","title":"...","description":"...",'
        '"daysFromArrival":1,"urgency":"critical","estimatedTime":"2-3 hours",'
        '"dueDate":"YYYY-MM-DD"}]}'
    )
    text = await _gen(prompt)
    return json.loads(_strip_fences(text))


async def next_onboarding_question(prior_qa: list[str], target_topic: str) -> str:
    context = ""
    if prior_qa:
        context = "Here is what we know about the user so far:\n" + "\n".join(prior_qa) + "\n\n"

    prompt = (
        "You are Roots, a warm AI companion for newcomers to Canada.\n"
        "You are currently onboarding a new user.\n"
        f"{context}"
        "It is your turn to ask the user a question. The core information you "
        f"need to gather is:\n\"{target_topic}\"\n\n"
        "Rephrase this core question into a single, naturally conversational "
        "and friendly message. You can briefly acknowledge their previous "
        "answer if it makes sense, but keep it short.\n"
        "DO NOT provide any advice yet, just ask the question.\n"
        "CRITICAL: Match the language the user has been writing in.\n"
        "Your entire response must be ONLY the question (1-2 sentences max)."
    )
    return (await _gen(prompt)).strip()


async def classify_interrupt(last_question: str, message: str) -> bool:
    prompt = (
        f"User is in onboarding. Last question asked: \"{last_question}\"\n\n"
        f"User said: \"{message}\"\n\n"
        "Is the user asking a question, expressing a concern, or asking for "
        "help that's UNRELATED to answering the specific question above? "
        "Respond with only 'YES' or 'NO'."
    )
    out = (await _gen(prompt)).strip().upper()
    return "YES" in out


async def analyse_document(image_b64: str, mime: str = "image/jpeg") -> dict[str, Any]:
    """DocumentAnalyst's actual Gemini call — multimodal extraction."""
    prompt = (
        "You are a strict data extraction assistant for Canadian immigrant "
        "documents.\n"
        "1. Identify the document type (passport, visa, PR card, SIN letter, "
        "study permit, work permit, driver's license, etc.).\n"
        "2. Extract key fields: name, document_number, expiry_date, "
        "issuing_country, plus any document-specific fields you see.\n"
        "3. NEVER output sensitive ID numbers in plain text. Always redact as "
        "\"[REDACTED XXXXX]\".\n"
        "4. Provide a confidence between 0 and 1.\n"
        "5. List intuitively missing companion documents in "
        "missing_companion_docs.\n\n"
        "Return ONLY JSON matching this shape:\n"
        '{"doc_type":"...","fields":{"name":"...","document_number":"[REDACTED XXXXX]",'
        '"expiry_date":"YYYY-MM-DD","issuing_country":"..."},'
        '"confidence":0.0,"missing_companion_docs":["..."]}'
    )

    body = {
        "contents": [
            {
                "parts": [
                    {"text": prompt},
                    {"inline_data": {"mime_type": mime, "data": image_b64}},
                ]
            }
        ],
        "generationConfig": {"response_mime_type": "application/json"},
    }
    text = await _post(body)
    return json.loads(_strip_fences(text))


async def _gen(prompt: str) -> str:
    return await _post({"contents": [{"parts": [{"text": prompt}]}]})


async def _post(body: dict[str, Any]) -> str:
    url = f"{_API_BASE}/{settings.gemini_model}:generateContent?key={_api_key()}"
    async with httpx.AsyncClient(timeout=60) as http:
        r = await http.post(url, json=body)
        r.raise_for_status()
        data = r.json()
    return data["candidates"][0]["content"]["parts"][0]["text"]


def fetch_image_base64(url: str) -> tuple[str, str]:
    """Synchronously download a Twilio media URL and return (b64, mime)."""
    import requests

    auth = (settings.twilio_account_sid, settings.twilio_auth_token)
    r = requests.get(url, auth=auth, timeout=30)
    r.raise_for_status()
    mime = r.headers.get("Content-Type", "image/jpeg").split(";")[0]
    return base64.b64encode(r.content).decode("ascii"), mime
